### ISSUE TITLE
[BUG] Assorted dataclass bugs from test_dataclasses

### ISSUE THREAD

**Issue by da-woods** [cython/cython#4956]
- [ ] ('TestCase', 'test_no_options'),  # `@dataclass()`
- [ ] ('TestCase', 'test_field_no_default'),  # field()
- [ ] ('TestCase', 'test_init_in_order'),  # field()
- [ ] ('TestCase', 'test_hash_field_rules'),  # compiler crash
- [ ] ('TestCase', 'test_class_var'),  # not sure but compiler crash
- [ ] ('TestCase', 'test_field_order'),  # invalid C code (__pyx_base?)
- [ ] ('TestCase', 'test_overwrite_fields_in_derived_class'),  # invalid C code (__pyx_base?)
- [ ] ('TestReplace', 'test_recursive_repr'),  # recursion error
- [ ] ('TestReplace', 'test_recursive_repr_two_attrs'),  # recursion error
- [ ] ('TestReplace', 'test_recursive_repr_misc_attrs'),  # recursion error
- [ ] ('TestReplace', 'test_recursive_repr_indirection'),  # recursion error
- [ ] ('TestReplace', 'test_recursive_repr_indirection_two'),  # recursion error
- [ ] ('TestCase', 'test_0_field_compare'),  # should return False
- [ ] ('TestCase', 'test_1_field_compare'),  # order=False is apparently ignored
- [ ] ('TestOrdering', 'test_no_order'),  # probably order=False being ignored
- [ ] ('TestRepr', 'test_no_repr'),  # turning off repr doesn't work
- [ ] ('TestCase', 'test_intermediate_non_dataclass'),  # issue with propagating through intermediate class
- [ ] ('TestCase', 'test_post_init'),  # init=False being ignored
- [ ] ('TestFrozen',),  # raises AttributeError, not FrozenInstanceError (may be hard to fix)
- [ ] ('TestReplace', 'test_frozen'),  # AttributeError not FrozenInstanceError
- [ ] ('TestCase', 'test_dataclasses_qualnames'),  # doesn't define `__setattr__` and just relies on Cython to enforce readonly properties
- [ ] ('TestCase', 'test_compare_subclasses'),  # wrong comparison
- [ ] ('TestCase', 'test_simple_compare'),  # wrong comparison
- [ ] ('TestEq', 'test_no_eq'),  # wrong comparison (probably eq=False being ignored)
- [ ] ('TestCase', 'test_field_named_self'),  # I think just an error in inspecting the signature
- [ ] ('TestCase', 'test_init_var_default_factory'),  # should be raising a compile error
- [ ] ('TestCase', 'test_init_var_no_default'),  # should be raising a compile error
- [ ] ('TestCase', 'test_init_var_with_default'),  # not sure...
- [ ] ('TestReplace', 'test_initvar_with_default_value'),  # needs investigating

A lot of these bugs are duplicates of course.

Possible bugs
- [ ]  # non-default argument 'z' follows default argument in dataclass `__init__` - this message looks right to me!
       ('TestCase', 'test_class_marker'),
- [ ] # cython.dataclasses.field parameter 'metadata' must be a literal value - possibly not something we can support?
      ('TestCase', 'test_field_metadata_custom_mapping'),
- [ ] ('TestCase', 'test_class_var_default_factory'),  # possibly to do with ClassVar being assigned a field
- [ ] ('TestCase', 'test_class_var_with_default'),  # possibly to do with ClassVar being assigned a field
- [ ] ('TestHash', 'test_unsafe_hash'),  # not sure if it's a bug or just a difference in how the hash is calculated
- [ ]  ('TestHash', 'test_1_field_hash'),  # not sure if it's a bug or just a difference in how the hash is calculated
- [ ] ('TestHash', 'test_0_field_hash'),  # not sure if it's a bug or just a difference in how the hash is calculated
- [ ] ('TestDescriptors',),  # mostly don't work - I think this may be a limitation of cdef classes but needs investigating

Also:
- [ ] Try to be clever about identifying classes that don't need to be cdef when translating the test (this'll get a lot more of the tests working)
- [ ] Translate some of the classes into strings to pass to Cython compile (where the test is written so that they use runtime options)

**Comment by scoder** [comment]
I just fixed a hash related bug while refactoring the code generation of `Dataclass.py` in https://github.com/cython/cython/commit/9cb28fb273b9bb9b2c08d4f6131e08f5a92f951c. Seems worth retrying if that still gives the same results.

**Comment by da-woods** [comment]
> I just fixed a hash related bug while refactoring the code generation of `Dataclass.py` in [9cb28fb](https://github.com/cython/cython/commit/9cb28fb273b9bb9b2c08d4f6131e08f5a92f951c). Seems worth retrying if that still gives the same results.

It doesn't seem to. To be honest I'm not sure that most of the hash issues are real bugs - I think it's just a consequence that CPython puts everything in a tuple and hashes that while Cython doesn't (because not every Cython field could go in a tuple). But it needs a bit more investigation before I'm definitely happy with that answer

**Comment by scoder** [comment]
> I think it's just a consequence that CPython puts everything in a tuple and hashes that while Cython doesn't

Yes, that's most likely the reason. Tuples have a specific base hash value. I don't really think that's worth emulating.
